3.8/graphs-custom-pregel.md

```diff
@@ -0,0 +1,1285 @@
+- Read messages which are sent to the vertex V in the previous superstep (S-1)
+- Send messages to other vertices that will be received in the next superstep (S+1)
+- Modify the state of the vertex V
+- Vote to Halt (mark a vertex V as "done".
+  V will be inactive in S+1, but it is possible to re-activate)
+
+Definition of a custom algorithm
+--------------------------------
+
+The format of a custom algorithm right now is based on a JSON object.
+
+### Algorithm skeleton
+
+```json
+{
+  "resultField": "<string>",
+  "maxGSS": "<number>",
+  "dataAccess": {
+    "writeVertex": "<program>",
+    "readVertex": "<array>",
+    "readEdge": "<array>"
+  },
+  "vertexAccumulators": "<object>",
+  "globalAccumulators": "<object>",
+  "customAccumulators": "<object>",
+  "phases": "<array>"
+}
+```
+
+### Algorithm parameters
+
+- **resultField** (string, _optional_): Name of the document attribute to store
+  the result in.
+  The vertex computation results will be in all vertices pointing to the given attribute.
+
+- **maxGSS** (number, _required_): The max amount of global supersteps
+  After the amount of max defined supersteps is reached, the Pregel execution will stop.
+
+- **dataAccess** (object, _optional_): Allows to define `writeVertex`,
+  `readVertex` and `readEdge`.
+
+  - **writeVertex**: A [program](#program) that is used to write the results
+    into vertices. If `writeVertex` is used, the `resultField` will be ignored.
+  - **readVertex**: An `array` that consists of `strings` and/or additional `arrays`
+    (that represents a path).
+    - `string`: Represents a single path at the top level which is **not** nested.
+    - `array of strings`: Represents a nested path
+  - **readEdge**: An `array` that consists of `strings` and/or additional `arrays`
+    (that represents a path).
+      - `string`: Represents a single path at the top level which is **not** nested.
+      - `array of strings`: Represents a nested path
+
+- **vertexAccumulators** (object, _optional_):
+  Definition of all used vertex accumulators.
+  <!-- TODO: Vertex Accumulators -->
+
+- **globalAccumulators** (object, _optional_):
+  Definition all used global accumulators.
+  Global Accumulators are able to access variables at shared global level.
+
+- **customAccumulators** (object, _optional_):
+  Definition of all used custom accumulators.
+
+- **phases** (array, _optional?_):
+  Array of a single or multiple [phase definitions](#phases).
+
+- **debug** (optional): See [Debugging](#debugging).
+
+Phases
+------
+
+Phases will run sequentially during your Pregel computation. The definition of
+multiple phases is allowed. Each phase requires instructions based on the
+operations you want to perform. The initialization program (1) will run a
+single time in the very first round. All upcoming rounds will execute the
+update program (2).
+
+The Pregel program execution will follow the order:
+
+Step 1: Initialization
+1. `onPreStep` (Coordinator)
```

Comment:
This get's a bit confusing by design.
It is unclear to the User That Coordinator and Conductor are identical.
In this part of the Documentation we claim onPreStep to be executed on the Coordinator.
Later we define Conductor specialities. 

I would suggest the following:

`onPreStep` (Conductor, executed on Coordinator instances)
`initProgram` ( Worker, executed on DBServer instances)